src/main/java/io/vertx/ext/web/handler/impl/FormLoginHandlerImpl.java

```diff
@@ -65,28 +93,41 @@ public void handle(RoutingContext context) {
         context.fail(400);
       } else {
         Session session = context.session();
-        if (session == null) {
-          context.fail(new NullPointerException("No session - did you forget to include a SessionHandler?"));
-        } else {
-          JsonObject authInfo = new JsonObject().put("username", username).put("password", password);
-          authProvider.authenticate(authInfo, res -> {
-            if (res.succeeded()) {
-              User user = res.result();
-              context.setUser(user);
+        JsonObject authInfo = new JsonObject().put("username", username).put("password", password);
+        authProvider.authenticate(authInfo, res -> {
+          if (res.succeeded()) {
+            User user = res.result();
+            context.setUser(user);
+            if (session != null) {
               String returnURL = session.remove(returnURLParam);
-              if (returnURL == null) {
-                context.fail(new IllegalStateException("Logged in OK, but no return URL"));
-              } else {
+              if (returnURL != null) {
                 // Now redirect back to the original url
-                req.response().putHeader("location", returnURL).setStatusCode(302).end();
+                doRedirect(req.response(), returnURL);
+                return;
               }
+            }
+            // Either no session or no return url
+            if (directLoggedInOKURL != null) {
+              // Redirect to the default logged in OK page - this would occur
+              // if the user logged in directly at this URL without being redirected here first from another
+              // url
+              doRedirect(req.response(), directLoggedInOKURL);
             } else {
-              context.fail(403);  // Failed login
+              // Just show a basic page
+              req.response().end(DEFAULT_DIRECT_LOGGED_IN_OK_PAGE);
```

Comment:
Shouldn't we also set the status to 403 ? Even if it's a browser getting the page, it's generally good to use the right HTTP status (the page will be interpreted as usual by the browser).